web-runtime/src/module/wit_instance.rs

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2020 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use super::wit_prelude::*;
+use super::IRecordType;
+use crate::{js_log, MResult};
+
+use marine_it_interfaces::MITInterfaces;
+use marine_it_interfaces::ITAstType;
+use wasmer_it::interpreter::wasm;
+use wasmer_it::interpreter::wasm::structures::{LocalImportIndex, Memory, TypedIndex};
+use crate::marine_js::{Instance as WasmerInstance, DynFunc};
+
+use std::collections::HashMap;
+use std::rc::Rc;
+
+pub type MRecordTypes = HashMap<u64, Rc<IRecordType>>;
+
+/// Contains all import and export functions that could be called from IT context by call-core.
+#[derive(Clone)]
+pub(super) struct ITInstance {
+    /// IT functions indexed by id.
+    funcs: HashMap<usize, WITFunction>,
+
+    /// IT memories.
+    memories: Vec<WITMemory>,
+
+    /// All record types that instance contains.
+    record_types_by_id: MRecordTypes,
+}
+
+impl ITInstance {
+    pub(super) fn new(wasmer_instance: &WasmerInstance, wit: &MITInterfaces<'_>) -> MResult<Self> {
+        let exports = Self::extract_raw_exports(wasmer_instance, wit)?;
+        let memories = Self::extract_memories(wasmer_instance);
+        js_log(&format!("executed Self::extract_memories"));
+
+        let funcs = exports;
+
+        let record_types_by_id = Self::extract_record_types(wit);
+        js_log(&format!("executed Self::extract_record_types"));
+
+        Ok(Self {
+            funcs,
+            memories,
+            record_types_by_id,
+        })
+    }
+
+    fn extract_raw_exports(
+        wasmer_instance: &WasmerInstance,
+        it: &MITInterfaces<'_>,
+    ) -> MResult<HashMap<usize, WITFunction>> {
+        let module_exports = &wasmer_instance.exports;
+
+        it.exports()
+            .enumerate()
+            .map(|(export_id, export)| {
+                let export_func = module_exports.get(export.name)?;
+                unsafe {
+                    // TODO: refactor this with new Wasmer API when it is ready
+                    // here it is safe because dyn func is never lives WITInstance
+                    let export_func =
```

Comment:
You don't need this transmuting, because it's needed in Marine in order to have a struct with self-reference, that contains instance of Wasmer and `DynFunc` that internally have a reference to Wasmer.